### PR TITLE
ISPN-5416 Grouping and aggregations with DSL based queries

### PR DESCRIPTION
**Pull request by anistor**
jira: https://issues.jboss.org/browse/ISPN-5416

### CODE REVIEW

**Comment by anistor** [comment]
The new HQL parser is released now, this should build fine.

**Comment by wburns** [comment]
I am guessing it is not easy and obviously wouldn't fall under this PR.  But the new distributed streams, which underlies non indexed query now has support for various aggregations built in.  Not sure if you can translate to those, but it should help non indexed query performance substantially when these are used.

https://github.com/infinispan/infinispan/blob/master/core/src/test/java/org/infinispan/stream/BaseStreamTest.java#L202 is a regular count and others are shown in the test class.

Grouping would be a bit more interesting though.

**Comment by anistor** [comment]
We'll make some incremental improvements to this feature and we'll surely get to do that too :) But let's get it in first for the beta release.

File: object-filter/src/main/java/org/infinispan/objectfilter/impl/BaseMatcher.java [hunk 1]
```diff
@@ -110,6 +114,12 @@ public FilterSubscription registerFilter(String jpaQuery, FilterCallback callbac
       return queryParser.parseQuery(jpaQuery, createFilterProcessingChain(namedParameters));
    }
 
+   private void disallowGroupingAndAggregations(FilterParsingResult<TypeMetadata> parsingResult) {
+      if (parsingResult.hasGroupingOrAggregations()) {
+         throw new ParsingException("Matcher and ObjectFilter do not allow grouping or aggregations");
```
Comment (tristantarrant) [inline object-filter/src/main/java/org/infinispan/objectfilter/impl/BaseMatcher.java]: Externalize this log

File: object-filter/src/main/java/org/infinispan/objectfilter/impl/hql/FilterRendererDelegate.java [hunk 2]
```diff
@@ -20,31 +22,80 @@
...
+               _aggregationType = PropertyPath.AggregationType.MIN;
+               break;
+            case MAX:
+               _aggregationType = PropertyPath.AggregationType.MAX;
+               break;
+            case COUNT:
+               _aggregationType = PropertyPath.AggregationType.COUNT;
+               break;
+            default:
+               throw new IllegalStateException("Aggregation " + aggregationType.name() + " is not supported");
```
Comment (tristantarrant) [inline object-filter/src/main/java/org/infinispan/objectfilter/impl/hql/FilterRendererDelegate.java]: Externalize log

File: object-filter/src/main/java/org/infinispan/objectfilter/impl/hql/RowPropertyHelper.java [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+      Class<?> propType = getColumnAccessor(propertyPath).getPropertyType();
+      if (propType.isEnum() || primitives.contains(propType)) {
+         return propType;
+      }
+      return null;
+   }
+
+   private ColumnMetadata getColumnAccessor(List<String> propertyPath) {
+      if (propertyPath.size() > 1) {
+         throw new IllegalStateException("Nested attributes are not supported");
```
Comment (tristantarrant) [inline object-filter/src/main/java/org/infinispan/objectfilter/impl/hql/RowPropertyHelper.java]: Externalize log

File: object-filter/src/main/java/org/infinispan/objectfilter/impl/hql/RowPropertyHelper.java [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+      }
+
+      String columnName = propertyPath.get(0);
+      for (RowPropertyHelper.ColumnMetadata c : rowMetadata.getColumns()) {
+         if (c.getColumnName().equals(columnName)) {
+            return c;
+         }
+      }
+
+      throw new IllegalArgumentException("Column not found : " + columnName);
```
Comment (tristantarrant) [inline object-filter/src/main/java/org/infinispan/objectfilter/impl/hql/RowPropertyHelper.java]: Externalize log

File: object-filter/src/main/java/org/infinispan/objectfilter/impl/hql/RowPropertyHelper.java [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+   }
+
+   @Override
+   public boolean hasProperty(String entityType, List<String> propertyPath) {
+      if (propertyPath.size() > 1) {
+         throw new IllegalStateException("Nested attributes are not supported");
```
Comment (tristantarrant) [inline object-filter/src/main/java/org/infinispan/objectfilter/impl/hql/RowPropertyHelper.java]: Externalize log

File: object-filter/src/main/java/org/infinispan/objectfilter/impl/hql/predicate/SingleEntityHavingQueryBuilderImpl.java [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+
+   public SingleEntityHavingQueryBuilderImpl(EntityNamesResolver entityNamesResolver, ObjectPropertyHelper propertyHelper) {
+      this.entityNamesResolver = entityNamesResolver;
+      this.propertyHelper = propertyHelper;
+   }
+
+   @Override
+   public void setEntityType(String entityType) {
+      if (entityNamesResolver.getClassFromName(entityType) == null) {
+         throw new IllegalStateException("Unknown entity name " + entityType);
```
Comment (tristantarrant) [inline object-filter/src/main/java/org/infinispan/objectfilter/impl/hql/predicate/SingleEntityHavingQueryBuilderImpl.java]: Externalize log

File: object-filter/src/main/java/org/infinispan/objectfilter/impl/hql/FilterRendererDelegate.java [hunk 7]
```diff
@@ -20,31 +22,80 @@
...
-         sortFields = new ArrayList<SortField>();
+         sortFields = new ArrayList<SortField>(5);
+      }
+      sortFields.add(new FilterParsingResult.SortFieldImpl(makePropertyPath(propertyPath), isAscending));
+   }
+
+   private PropertyPath makePropertyPath(org.hibernate.hql.ast.origin.hql.resolve.path.PropertyPath propertyPath) {
+      AggregationPropertyPath.Type aggregationType = propertyPath instanceof AggregationPropertyPath ? ((AggregationPropertyPath) propertyPath).getType() : null;
+      PropertyPath.AggregationType _aggregationType = null;
+      if (aggregationType != null) {
```
Comment (tristantarrant) [inline object-filter/src/main/java/org/infinispan/objectfilter/impl/hql/FilterRendererDelegate.java]: You already have this code https://github.com/infinispan/infinispan/pull/3591/files#diff-0bccd6825414ee3a8e5e98e6e38c4688R143
Put it somewhere reusable

File: query-dsl/src/main/java/org/infinispan/query/dsl/QueryBuilder.java [hunk 8]
```diff
@@ -8,10 +8,22 @@
...
 
+   QueryBuilder select(Expression... projection);
+
+   QueryBuilder select(String... attributePath);
+
+   /**
+    * @see #select(String... attributePath)
+    */
+   @Deprecated
    QueryBuilder setProjection(String... attributePath);
```
Comment (tristantarrant) [inline query-dsl/src/main/java/org/infinispan/query/dsl/QueryBuilder.java]: You should add this to the upgrading.adoc file

File: query-dsl/src/main/java/org/infinispan/query/dsl/impl/BaseQueryBuilder.java [hunk 9]
```diff
@@ -51,29 +58,74 @@ protected String getRootTypeName() {
...
+
    protected List<SortCriteria> getSortCriteria() {
       return sortCriteria;
    }
 
    @Override
-   public QueryBuilder<T> setProjection(String... projection) {
+   public QueryBuilder<T> select(String... attributePath) {
+      if (attributePath == null || attributePath.length == 0) {
+         throw new IllegalArgumentException("Projection cannot be null or empty");
```
Comment (tristantarrant) [inline query-dsl/src/main/java/org/infinispan/query/dsl/impl/BaseQueryBuilder.java]: Externalize log

File: query-dsl/src/main/java/org/infinispan/query/dsl/impl/BaseQueryBuilder.java [hunk 10]
```diff
@@ -51,29 +58,74 @@ protected String getRootTypeName() {
...
+      for (int i = 0 ; i < attributePath.length; i++) {
+         projection[i] = Expression.property(attributePath[i]);
+      }
+      return select(projection);
+   }
+
+   @Override
+   public QueryBuilder<T> select(Expression... projection) {
+      if (projection == null || projection.length == 0) {
+         throw new IllegalArgumentException("Projection cannot be null or empty");
```
Comment (tristantarrant) [inline query-dsl/src/main/java/org/infinispan/query/dsl/impl/BaseQueryBuilder.java]: Externalize log

File: query-dsl/src/main/java/org/infinispan/query/dsl/impl/BaseQueryBuilder.java [hunk 11]
```diff
@@ -51,29 +58,74 @@ protected String getRootTypeName() {
...
+      }
+      if (this.projection != null) {
+         throw new IllegalStateException("Projection can be specified only once");
```
Comment (tristantarrant) [inline query-dsl/src/main/java/org/infinispan/query/dsl/impl/BaseQueryBuilder.java]: Externalize log

File: query-dsl/src/main/java/org/infinispan/query/dsl/impl/BaseQueryBuilder.java [hunk 12]
```diff
@@ -51,29 +58,74 @@ protected String getRootTypeName() {
...
+   }
+
+   protected Expression[] getProjection() {
       return projection;
    }
 
    @Override
+   public QueryBuilder groupBy(String... groupBy) {
+      if (groupBy == null || groupBy.length == 0) {
+         throw new IllegalArgumentException("Grouping cannot be null or empty");
```
Comment (tristantarrant) [inline query-dsl/src/main/java/org/infinispan/query/dsl/impl/BaseQueryBuilder.java]: Externalize log

File: query-dsl/src/main/java/org/infinispan/query/dsl/impl/BaseQueryBuilder.java [hunk 13]
```diff
@@ -51,29 +58,74 @@ protected String getRootTypeName() {
...
+      }
+      if (this.groupBy != null) {
+         throw new IllegalStateException("Grouping can be specified only once");
```
Comment (tristantarrant) [inline query-dsl/src/main/java/org/infinispan/query/dsl/impl/BaseQueryBuilder.java]: Externalize log

File: query-dsl/src/main/java/org/infinispan/query/dsl/impl/BaseQueryBuilder.java [hunk 14]
```diff
@@ -51,29 +58,74 @@ protected String getRootTypeName() {
...
+   }
+
+   protected String[] getGroupBy() {
+      return groupBy;
+   }
+
+   @Override
    public QueryBuilder<T> startOffset(long startOffset) {
       if (startOffset < 0) {
          throw new IllegalArgumentException("startOffset cannot be less than 0");
```
Comment (tristantarrant) [inline query-dsl/src/main/java/org/infinispan/query/dsl/impl/BaseQueryBuilder.java]: Externalize log

**Comment by tristantarrant** [comment]
@anistor you can ignore all the externalize log messages for now :) they can be fixed in a later iteration. The rest of the code looks good to me

**Comment by anistor** [comment]
Code duplication is removed now. Let's wait for CI...

File: integrationtests/all-embedded-query-it/src/test/java/org/infinispan/all/embeddedquery/QueryDslConditionsTest.java [hunk 15]
```diff
@@ -1724,6 +1786,55 @@ public void testUnorderedPagination8() throws Exception {
...
+   @Test(expected = ParsingException.class)
+   public void testGroupBy3() throws Exception {
+      QueryFactory qf = getQueryFactory();
+
+      Query q = qf.from(getModelFactory().getUserImplClass())
+            .select("name")
+            .groupBy("name")
+            .orderBy("surname", SortOrder.ASC)
+            .build();
+
```
Comment (gustavonalle) [inline integrationtests/all-embedded-query-it/src/test/java/org/infinispan/all/embeddedquery/QueryDslConditionsTest.java]: What about doing aggregation in a filtered query? Would something conceptually similar to ```SELECT name FROM person WHERE name.startsWith("a") GROUP BY name``` work? It seems there is no test for it

**Comment by gustavonalle** [comment]
Just one comment, looks very nice!  :+1:

File: integrationtests/all-embedded-query-it/src/test/java/org/infinispan/all/embeddedquery/QueryDslConditionsTest.java (hunk 15 above)
Comment (anistor) [inline integrationtests/all-embedded-query-it/src/test/java/org/infinispan/all/embeddedquery/QueryDslConditionsTest.java]: That should work. I think grouping + where is tested in QueryEngineTest. I'll add a new test for that in QueryDslConditionsTest too.

**Comment by wburns** [comment]
> We'll make some incremental improvements to this feature and we'll surely get to do that too :) But let's get it in first for the beta release.

@anistor Yeah that is my hope, I just wanted to make you aware is all, which is what I was trying to get at when I said this is not for this PR.  But this is looks nice! :)

File: object-filter/src/main/java/org/infinispan/objectfilter/impl/aggregation/AvgAccumulator.java [hunk 16]
```diff
@@ -0,0 +1,51 @@
+package org.infinispan.objectfilter.impl.aggregation;
+
+/**
+ * @author anistor@redhat.com
+ * @since 8.0
+ */
+public final class AvgAccumulator extends FieldAccumulator {
+
+   private static class Avg {
```
Comment (wburns) [inline object-filter/src/main/java/org/infinispan/objectfilter/impl/aggregation/AvgAccumulator.java]: Have you looked at using https://docs.oracle.com/javase/8/docs/api/java/util/DoubleSummaryStatistics.html ?  It does some special handling for averages that is probably more advanced then a simple division.

File: object-filter/src/main/java/org/infinispan/objectfilter/impl/aggregation/CountAccumulator.java [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+   }
+
+   @Override
+   public void init(Object[] row) {
+      row[pos] = 1;
+   }
+
+   @Override
+   public void update(Object[] srcRow, Object[] destRow) {
+      //todo [anistor] should we count nulls too?
```
Comment (wburns) [inline object-filter/src/main/java/org/infinispan/objectfilter/impl/aggregation/CountAccumulator.java]: I believe we should.

File: query-dsl/src/main/java/org/infinispan/query/dsl/QueryBuilder.java [hunk 18]
```diff
@@ -8,10 +8,22 @@
  */
 public interface QueryBuilder<Q extends Query> extends FilterConditionBeginContext {
 
+   QueryBuilder orderBy(Expression expression, SortOrder sortOrder);
```
Comment (wburns) [inline query-dsl/src/main/java/org/infinispan/query/dsl/QueryBuilder.java]: Do we want orderBy using a Comparator as well?

File: query-dsl/src/main/java/org/infinispan/query/dsl/impl/BaseQueryBuilder.java (hunk 9 above)
Comment (wburns) [inline query-dsl/src/main/java/org/infinispan/query/dsl/impl/BaseQueryBuilder.java]: Some of these seem like programming errors, which I wouldn't expect an externalized log or do we want to externalize those cases as well?

File: query-dsl/src/main/java/org/infinispan/query/dsl/QueryBuilder.java (hunk 18 above)
Comment (anistor) [inline query-dsl/src/main/java/org/infinispan/query/dsl/QueryBuilder.java]: I would say no. The values themselves are Comparable. The query language is supposed to be language neutral, useable in both local and remote mode. The Comparator is a java thing.

File: object-filter/src/main/java/org/infinispan/objectfilter/impl/aggregation/CountAccumulator.java (hunk 17 above)
Comment (anistor) [inline object-filter/src/main/java/org/infinispan/objectfilter/impl/aggregation/CountAccumulator.java]: +1

File: object-filter/src/main/java/org/infinispan/objectfilter/impl/aggregation/SumAccumulator.java [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+
+   @Override
+   public void update(Object[] srcRow, Object[] destRow) {
+      Number value = (Number) srcRow[pos];
+      if (value != null) {
+         Number sum = (Number) destRow[pos];
+         if (sum == null) {
+            destRow[pos] = value;
+         } else {
+            destRow[pos] = sum.intValue() + value.intValue(); //todo [anistor] this handles integers only!
```
Comment (wburns) [inline object-filter/src/main/java/org/infinispan/objectfilter/impl/aggregation/SumAccumulator.java]: I would say we should use a long and double respectively.

File: query-dsl/src/main/java/org/infinispan/query/dsl/QueryBuilder.java (hunk 18 above)
Comment (wburns) [inline query-dsl/src/main/java/org/infinispan/query/dsl/QueryBuilder.java]: Sounds fine to me, just wanted to check.

File: object-filter/src/main/java/org/infinispan/objectfilter/impl/aggregation/AvgAccumulator.java (hunk 16 above)
Comment (anistor) [inline object-filter/src/main/java/org/infinispan/objectfilter/impl/aggregation/AvgAccumulator.java]: No, but I see what you mean. Handling those pesky not-numbers... I'll address this in another PR.

File: object-filter/src/main/java/org/infinispan/objectfilter/impl/aggregation/SumAccumulator.java (hunk 19 above)
Comment (anistor) [inline object-filter/src/main/java/org/infinispan/objectfilter/impl/aggregation/SumAccumulator.java]: Actually I think we need to have IntegerSumAccumulator, LongSumAccumulator, DoubleSumAccumulator... etc. Same for avg, min, max.

File: query-dsl/src/main/java/org/infinispan/query/dsl/impl/BaseQueryBuilder.java (hunk 9 above)
Comment (anistor) [inline query-dsl/src/main/java/org/infinispan/query/dsl/impl/BaseQueryBuilder.java]: Some of these are triggered by user's programming errors, some are caused by our programming errors. Which ones do we i18n?

File: object-filter/src/main/java/org/infinispan/objectfilter/impl/aggregation/SumAccumulator.java (hunk 19 above)
Comment (wburns) [inline object-filter/src/main/java/org/infinispan/objectfilter/impl/aggregation/SumAccumulator.java]: Alright depends on the contract you want is all.  Some places allow int sums overflowing into a long value, either way is fine with me as long as we define it.

Comment (anistor) [inline object-filter/src/main/java/org/infinispan/objectfilter/impl/aggregation/SumAccumulator.java]: I've updated the accumulator to correctly compute sums based on the type of the property. Byte and short are promoted to int as it normally happens in java, the others remain unchanged.

File: object-filter/src/main/java/org/infinispan/objectfilter/impl/aggregation/AvgAccumulator.java (hunk 16 above)
Comment (anistor) [inline object-filter/src/main/java/org/infinispan/objectfilter/impl/aggregation/AvgAccumulator.java]: I have not fixed it here yet, but I added a jira: https://issues.jboss.org/browse/ISPN-5629

**Comment by wburns** [comment]
Pulling..

**Comment by wburns** [comment]
Integrated into master, thanks @anistor !

**Pull request closed without merging**